ultra/ultra/scenarios/common/begin_time_init_funcs.py

```diff
@@ -71,16 +76,39 @@ def burst_begin_time_init_func(
             vehicle_spawn_time += time_so_far
             _begin_times.extend(vehicle_spawn_time)
 
-            time_to_next_cluster = np.random.uniform(*time_between_cluster)
-            time_so_far += time_to_next_cluster
+            speed = "".join(filter(lambda i: i.isdigit(), speed))
+            if density == "high-density":
+                if int(speed) == 50:
+                    variable_time_between_cluster = (
+                        time_between_cluster[0] + speed_offset + (num_lanes / 2),
+                        time_between_cluster[1] + speed_offset + (num_lanes / 2),
+                    )
+                elif int(speed) == 70:
+                    variable_time_between_cluster = (
+                        time_between_cluster[0],
+                        time_between_cluster[1],
+                    )
+                elif int(speed) == 100:
+                    variable_time_between_cluster = (
+                        time_between_cluster[0] - speed_offset - (num_lanes / 2),
+                        time_between_cluster[1] - speed_offset - (num_lanes / 2),
+                    )
+                else:
+                    raise ValueError(f"Unsupported speed: {speed}.")
+
+                time_to_next_cluster = np.random.uniform(*variable_time_between_cluster)
+            else:
+                time_to_next_cluster = np.random.uniform(*time_between_cluster)
```

Comment:
I see, so you design the traffic density subject to the speed.  I wonder if it might be an issue if we introduce more speeds. We could have a piecewise linear function as a function of the speed without the if condition. Something like :

```
if speed >= 70
  speed_factor = (speed - 70)/30
elseif speed < 70 
  speed_factor = (speed-70)/20
```

and then have each component of variable_time_between_cluster use the speed_factor, like so `time_between_cluser[0] + (speed_factor)*(speed_offset - num_lanes/2)`.

We probably won't add more speeds, but it might be good to "future proof"

Reply:
@AlexLewandowski Christian and I had a discussion on this as well [https://github.com/huawei-noah/SMARTS/pull/783#discussion_r625777493](url). The current setup takes only into account of the three speeds that we have. In terms of the function of speed that governs the began times for variable speed still has to reconcile with. I like the idea of the piecewise linear function. We can discuss this in the ULTRA meeting on wednesday